File: 0345-reverse-vowels-of-a-string /0345-reverse-vowels-of-a-string.java
class Solution {
    public String reverseVowels(String s) {
        // Create a boolean array to mark vowels (both uppercase and lowercase)
        boolean[] vowels = new boolean[128];
        for (char c : "aeiouAEIOU".toCharArray()) {
            vowels[c] = true;
        }

        // Convert input string to character array
        char[] cs = s.toCharArray();
        int i = 0, j = cs.length - 1;

        // Two-pointer approach to find and swap vowels
        while (i < j) {
            // Move 'i' forward until a vowel is found
            while (i < j && !vowels[cs[i]]) {
                ++i;
            }
            // Move 'j' backward until a vowel is found
            while (i < j && !vowels[cs[j]]) {
                --j;
            }
            // Swap vowels at indices i and j
            if (i < j) {
                char t = cs[i];
                cs[i] = cs[j];
                cs[j] = t;
                ++i;
                --j;
            }
        }
        // Return the modified string
        return String.valueOf(cs);
    }
}

File: 0345-reverse-vowels-of-a-string /0345-reverse-vowels-of-a-string.py

class Solution:
    def reverseVowels(self, s: str) -> str:
        # Define vowels and initialize two pointers i (start) and j (end)
        vowels = "aeiouAEIOU"
        i, j = 0, len(s) - 1
        # Convert string to list for easy swapping
        cs = list(s)
        
        # Loop until the two pointers meet
        while i < j:
            # Move the start pointer forward until a vowel is found
            while i < j and cs[i] not in vowels:
                i += 1
            # Move the end pointer backward until a vowel is found
            while i < j and cs[j] not in vowels:
                j -= 1
            # Swap the vowels at i and j
            if i < j:
                cs[i], cs[j] = cs[j], cs[i]
                i, j = i + 1, j - 1
        
        # Join the list back into a string and return the result
        return "".join(cs)
